playbooks/app_dev/properties-management.adoc

```diff
@@ -0,0 +1,109 @@
+When an application is promoted through environments (such a DEV, IT, QA) some of its configuration properties need to change, for example the connection string to a database or the URL of a called web service. +
+How an application expects to read its configurations is completely application-dependent. That said, over the course of several projects we have seen some patterns emerge that we have found to be successful. There is no better or worse approach - it is the responsibility of the pipeline designer to choose the best approach for a given context. +
+This article focuses on environment-dependent properties, but the same approaches could be potentially used for all properties, whether or not they are environment-dependent. Credential management is out of scope of this article.  +
+Here are four approaches that that cover the most common use cases. 
+
+== Using Environment Variables
+
+Properties are passed by OpenShift as environment variables. The following is a fragment of how a such a pod template (or higher structures such as replication controllers and deployment configs) would be configured: +
+----
+          env: 
+            - name: MY_EXTERNAL_ENDPOINT 
+              value: http://xxx.yyy.zzz[http://xxx.yyy.zzz]
+----
+Your application needs to be trained to read its properties this way. If you are using Spring, there are several way of achieving this. Here is one example: +
+----
+@Value("#{systemEnvironment['MY_EXTERNAL_ENDPOINT']}")
+private String myExternalEndpoint;
+----
+
+Things to keep in mind when adopting this approach are that if the value of a property changes, your CI/CD workflow needs to be able to change it in the pod template and if a property is added or removed, your CI/CD workflow needs to be able to manage this event appropriately. +
+In general, this approach does not scale well when you have more than a dozen of properties.
+
+== Using One Environment Variable to Determine the Environment
+
+In this approach, we use an environment variable to determine in which environment the application is being deployed. The pod template would look like the following: +
+----
+          env: 
+            - name: ENV 
+              value: {IT|QA|PROD}
+----
+The docker image of your application will need to contain all the properties for each possible environment and to be able to select the right configuration file based on the ENV environment variable. +
+Using Spring, this can be elegantly achieved using profiles. Here is an example: +
+You will have several config classes, one per environment with the following annotations: +
+----
+@Configuration  
+@PropertySource("file:///<well-known-location>/application-dev.properties")  
+@Profile("dev")
+----
+Then this pod configuration to activate the right profile: +
+----
+          env: 
+            - name: SPRING_ACTIVE_PROFILE 
+              value: dev
+----
+Note: in this case we don’t even need the ENV variable +
```

Comment:
My apologies. This should have been `NOTE:` to trigger the callout formatting. Also, Can we capitalize the beginning of the sentence and remove the first line break?

Reply:
done